### DISCUSSION TITLE
php@7.2 has been disabled because it is deprecated upstream!

### DISCUSSION THREAD

**Discussion by leqq00** [Homebrew/discussions#2669] [Everyday usage] [answered]
php@7.2 has been disabled because it is deprecated upstream!

这个是什么情况？ 无法安装 老版本的PHP吗

**Comment by SMillerDev** [comment, marked as answer]
Yes, see https://www.php.net/eol.php

**Reply by brettvitaz** [reply]
This is not a reason to remove it. Many people rely on it. We're not asking for updates, we're asking for a reasonable and secure way to install it.

**Reply by SMillerDev** [reply]
Which we can't grant because it does not get maintenance or security fixes anymore. So we let people know that this was going to happen and then disabled it.

**Reply by lukeholder** [reply]
Thanks for being a maintainer, but totally disagree with this move. How can people upgrade their projects on a new computer if they cant use the same tools to switch php versions from 7.2 (run tests etc) to 8.1.

**Reply by SMillerDev** [reply]
That's fine, you don't have to agree. The way to get an unsupported version installed would be to maintain your own copy of the software in your own tap. Then nobody can restrict what you install.

https://docs.brew.sh/Manpage#extract---version---force-formula-tap and https://docs.brew.sh/How-to-Create-and-Maintain-a-Tap should help you with that.

**Reply by gromgit** [reply]
Or, as already mentioned before, tap (pun intended) on others' work: https://github.com/shivammathur/homebrew-php

**Reply by brettvitaz** [reply]
> Which we can't grant because it does not get maintenance or security fixes anymore

This is absolutely not a reason to remove it. I am not sure why you would think that excuse holds any weight. PHP 7 may not receive updates, but it is not your responsibility to determine what level of risk we incur on our own computers, especially with a valuable and highly relied upon application such as this. It is far riskier for us to install unknown taps to make up the loss when it's arbitrarily decided that we cannot reinstall an application that was working for years prior.

**Reply by MuneebRabaney** [reply]
And on a MONDAY - Boo hooo!

I feel both ends, however, @SMillerDev - douche move man, people install the entire universe with Node. I mean let's face it, a language is only as safe as the hacker sitting, feeding instructions to the machine. 

I have to admit, Yes we want up-to-date software, however, we update the core i.e "PHP" and all of our other packages break. I'd rather have the vulnerability and go in for a fix myself rather than 15 other packages that now broke because of something you just decided to take away.

If you want the community to respect you and this "Personal Home Page" language, you have to show care for others not just be all arrogant and like "Oh well tough titties, It's not my problem! Go make your own tap you, you, youuuuuu PHP 7 wanting a-hole"! Heck if my nose wasn't shoved up NodeJS's  and typescript butt Id have opted to maintain this project so that the world can live to see another day of our good ol' friend who they love to hate,  Mr PHP 7. I mean it's not like I'm asking for 5 or 4! Sigh


I'm done, anger out. Love you all - now back to find a workaround, Thanks for the eeeerrrrrrr, Therapy?
:)

**Reply by SMillerDev** [reply]
> I'd rather have the vulnerability and go in for a fix myself rather than 15 other packages that now broke because of something you just decided to take away.

If you want to manage it yourself, maybe using a package manager to manage your packages is the wrong approach for you.

Homebrew has clearly defined rules about what is supported, if you don't like those there are multiple ways to override them. If you don't like how Homebrew does things and you don't want to do it yourself I suggest [this](https://mikemcquaid.com/2018/03/19/open-source-maintainers-owe-you-nothing/)

**Reply by SMillerDev** [reply]
It's relevant because it's by the project lead of homebrew, therefore reinforcing my position that if you don't want to maintain something, homebrew doesn't have to either because it's license literally says it doesn't.

**Reply by SMillerDev** [reply]
>  The maintainers provide no assurances that the software will ever work for any user or use case (even documented ones)

Here you go, there was no assurance that the software would work and you got lucky that it did until it didn't.

> The maintainers are never liable for any problems caused by any use of the software (including damages that require you to pay for repairs)

Hey look, you used Homebrew and it caused the problem that you can't install a version of PHP that PHP no longer supports. Homebrew is not liable for that.

> We only ask that Homebrew does not remove important, functional, working, stable packages.

If you want to pin your software stack to a specific version, Homebrew (or any rolling release) is not the tool I'd recommend using. Docker does this way better. 

We can argue what it means to be a maintainer for an open source project for a while, but in the end the simple truth is this:
- Homebrew has rules about versions, we followed them here and nothing about that will change.

**Reply by MuneebRabaney** [reply]
ok; ok; ok Guys 😓, Yes @SMillerDev I agree with you; I'll take a knee on this one as I was wrong. I appreciate what You and other maintainers are doing to keep things going with great GIRTH, so be a fella and please accept my apologies as that rant was out of line. sigh. I need a yellow rubber duckie for my rants, I was a bit ticked off this morning. 

We all have those crap Mondays where you worked all weekend trying to make something awesome; and then;  there's this that one annoying thing you always use to be able to do with ease - however, now, you can't, and that's all she wrote...

(i.e the switching between PHP versions using homebrew), now you have to use brain power to compute another way. I love homebrew and have been using it for a while, in my opinion, It's the best, super easy and simple to use. Also thanks for the blog links and helpful comments. I managed to fix my issue and ship my code, easy peasy no stress; no; no; no :). And on the plus; 

I now have a way forward should I ever be in a similar situation with other software. If there's a way I can contribute I'll submit a PR.  

You're doing what's best for us and that sits well in my books 
Kudos.

**Comment by jjonline** [comment]
`monterey12.1`

Because some old projects still need php7.2 to be used, is there an alternative installation method?

**Reply by SMillerDev** [reply]
You can try people's personal tap, like https://github.com/shivammathur/homebrew-php

**Reply by senguttuvang** [reply]
I have inherited a project from another team which used 7.2. It isn't advisable to start with 7.3 or 8.0. I need to first run the app as it is, then slowly migrate to the newer version. The new developer can't be held responsible for what is been maintained by the previous developer/team. You may discourage the usage but do not block it. Let the developer decide which version to use based on several other factors (business priorities, backward compatibility for related package, etc). 

Yes, I went with the personal tap mentioned. Thank you!

**Reply by SMillerDev** [reply]
We're not just arbitrarily restricting PHP versions as you seem to imply. https://www.php.net/eol.php has the supported PHP versions and Homebrew/core is simply following the (PHP) projects direction in this. If you want longer support for PHP versions you can always write to the internals mailinglist requesting such support.

**Reply by keeev** [reply]
> You can try people's personal tap, like https://github.com/shivammathur/homebrew-php

thanks, this hint helped me running my Laravel Project again which depended on 7.3 while others were running on 7.4. 🙌🏼

In case anyone is looking for a solution after updating brew and not being able to run `php artisan ...` and resulting in: 

```
 dyld: Library not loaded: /usr/local/opt/openldap/lib/libldap-2.4.2.dylib
  Referenced from: /usr/local/bin/php
  Reason: image not found
```

this might be working for you too.

**Reply by henrypenny** [reply]
I'm glad I don't have to make my own decisions about which php versions I need to do my job.

**Comment by carcinocron** [comment]
Yeah, I have no idea why homebrew randomly uninstalled php without my consent. It's really fucking annoying. This has been extremely disruptive and anti-user. Normal software would give you a yellow warning message. This is some bullshit I'd expect from windows. It's not brew's job to FORCE USERS to stop developing software on "days old" versions of PHP. Now it's going to take EVEN LONGER for me to upgrade PHP because I had to do this work-around bullshit.

**Reply by carlocab** [reply]
A formula being disabled does not forcibly uninstall it from your system.

There are still users with installations of Homebrew `python@2` around, even if that formula is long gone by now.

**Reply by SMillerDev** [reply]
On top of the earlier comment:

> Normal software would give you a yellow warning message.

Homebrew did for a year when PHP 7.2 was in security-only mode

**Reply by carcinocron** [reply]
> A formula being disabled does not forcibly uninstall it from your system.

I wouldn't be here if brew did not forcibly uninstall it from my system.

**Reply by gromgit** [reply]
> I wouldn't be here if brew did not forcibly uninstall it from my system.

There are only two situations I can think of where you'd lose your PHP 7.2 installation:
1. You had a 7.2-era `php` formula rather than a versioned `php@7.2` formula, i.e. you hadn't upgraded your `php` in quite a while. (The former would absolutely be upgraded, the latter wouldn't be touched at all.)
2. You uninstalled Homebrew (which would remove all formulae), then reinstalled it.

If you still remember the circumstances under which your PHP 7.2 disappeared, and it's not one of the above, please post details.

**Reply by carcinocron** [reply]
I updated my version of crystal, which is completely unrelated to php, then php was broken on my machine.

**Reply by carlocab** [reply]
They're not completely unrelated. `crystal` and `php@7.2` share a number of common dependencies, and so are connected via the dependency graph:
```
❯ brew deps crystal php@7.2
ca-certificates
gmp
openssl@1.1
readline
sqlite
```
See `brew deps --graph crystal php@7.2`.

**Reply by odahcam** [reply]
@SMillerDev don't you understand you are breaking the fuck out of old software just because it's DEPRECATED? PHP didn't erase version 7.2 from the internet, come on guys, you can't be serious!

**Reply by SMillerDev** [reply]
Do you understand you're complaining to volunteers who don't owe you anything? If homebrew volunteers collectively agree to stop shipping php altogether they can, I don't owe you any support.

**Reply by henrypenny** [reply]
@SMillerDev demanding that volunteers actively 'support' php7.2 is obviously not cool. But it appears that one of those volunteers went out of their way to prevent 7.2 from being used. Why not just leave 7.2 alone and allow people to continue as before, with zero support offered if something goes wrong?

Having said all that, a solution was offered by @SMillerDev earlier that worked just fine for me: 
https://github.com/shivammathur/homebrew-php
So @odahcam, perhaps you should try that alternative, be grateful that Homebrew exists at all and get on with your life.

**Reply by SMillerDev** [reply]
> Why not just leave 7.2 alone and allow people to continue as before, with zero support offered if something goes wrong?

Because telling people that make issues about it that there is no support is also a form of support. The only way to provide no support is for the formula not to exist.

**Reply by carcinocron** [reply]
surely telling people you don't support php 7.2 is easier than telling people you have no problem remotely deleting php 7.2 from their computers...?

**Reply by SMillerDev** [reply]
It is, that's why we don't do that.

**Reply by carcinocron** [reply]
but this entire thread exists because you do delete php 7.2 from people's computers without their consent.

**Reply by carlocab** [reply]
You're the only one here who claims that has happened to them. You've definitely misunderstood the consequences of formulae getting disabled.

I don't know why `php@7.2` was uninstalled from your Mac, but it wasn't because the formula was disabled.

Perhaps you hit a bug, or perhaps you removed it yourself, probably accidentally. If it's a bug, though, then it's the first I'm hearing of it.

**Reply by lukeholder** [reply]
> You're the only one here who claims that has happened to them

Its happening to lots of us.

Tools like Laravel Valet directly document switching between homebrew installed PHP versions between 7.2 -> 8.1 etc.

https://laravel.com/docs/9.x/valet#php-versions

**Reply by carlocab** [reply]
> > You're the only one here who claims that has happened to them
> 
> 
> 
> Its happening to lots of us.

Please provide reproduction steps.

**Reply by eshe-huli** [reply]
I was maintaining from 7.1 to 7.4 in my computer after brew upgrade I first lost 7.1 after another I just realize 7.2 is missing so yes you do remove them from my computer without telling me.

**Reply by SMillerDev** [reply]
That's interesting, how were you maintaining them in your computer? Since brew should not have access to non-brew files.

**Reply by odahcam** [reply]
@henrypenny @SMillerDev the thing is not providing support is just making no changes to anything, you simply do nothing, that's the easiest thing. Now the moment you **delete** something you are changing stuff and that's what I can't understand. You simply have no obligation with anyone as I don't in my open source projects either, but I assure you that if I go there and delete some API **I don't like and think it's bad** that some folks make use, they will be **very** pissed off. If you just think about what causes less impact: how many hours you spent explaining why you deleted the formula instead of leaving it deprecated with a message that does that already and also gives the users freedom of choice. You do realize that lots of CI jobs makes use of homebrew to install dependecies just broke because of that? I understand it's your decision and you own no one nothing, but it's a very poor decision in my point of view. I can't trust brew official formulas for LTS software I write anymore, but I'm not complaining anymore, I'm jsut giving my opinion which you should have no obligation to care at all too.

**Reply by SMillerDev** [reply]
> the thing is not providing support is just making no changes to anything, you simply do nothing, that's the easiest thing.

Except that would mean that every update to the python or pcre formula for example would recompile and test an old php that is not likely to be compatible. Or, if we block the CI tests it would compile on your own device and break. (Which would lead to users asking for support)

> I can't trust brew official formulas for LTS software I write anymore

As a rolling release system that is not a recommended approach to begin with. Brew is pretty transparant in its goal to always strive for the latest version.
